### DISCUSSION TITLE
Crop image but still mapping annotation

### DISCUSSION THREAD

**Discussion by winter2897** [ashnair1/COCO-Assistant#35] [Q&A] [answered]
Hi @ashnair1 ,
I want to crop the large image into several small images and retain the annotation mapping. Can you give me advice?
![a](https://user-images.githubusercontent.com/22476495/111749060-e06a5f80-88c3-11eb-9013-951d6a126dea.png)

**Comment by ashnair1** [comment]
I'm not sure what you mean by annotation mapping. Is the annotation made at the large image level? So for example, you have one large image and one annotation file and you want to know how to tile it in order to make it a COCO style dataset?

**Comment by winter2897** [comment]
I already have the label file of the large image. I want it when I cut it into small pictures. These small pictures are automatically labeled in the COCO format. @ashnair1

**Comment by ashnair1** [comment]
What format is the label file? 

It seems you're working on geospatial imagery so I'm assuming your image is a GeoTIFF and label file is a GeoJSON/Shapefile. Is that correct?

**Comment by winter2897** [comment]
I'm working on large image sizes of solar panels. I label them with `labelme` and then use the code `labelme2coco.py`
![Screenshot from 2021-03-19 17-37-22](https://user-images.githubusercontent.com/22476495/111767865-cb003000-88d9-11eb-92ba-57ebdc672825.png)
 to convert the `.json` files to COCO format which is a single file` train.json`.

**Comment by winter2897** [comment]
Since the size image is too large, I want to cut it into small sheets. These small photos, after being cut, must automatically be labeled in the `COCO format`.

**Reply by ashnair1** [reply]
@winter2897 
Unfortunately I don't know of any tools that will do what you want directly. 

The most obvious solution would be:
1. Crop images to specific size (for example 256 x 256)
2. Label the image tiles using `labelme` and convert them to COCO format using `label2coco.py`

This means that you will need to do the annotation again but depending on your situation it might be easier.

Alternatively, if you don't want to do the annotation, you will need to write a script to modify the annotation

One way you could do this:
1. Tile images to specific size (for example 256 x 256)
2. Convert your large COCO annotation into a segmentation mask and tile that to size 256 x 256
3. Now try converting segmentation mask to coco annotation

**Reply by ashnair1** [reply]
You could refer [det2seg.py](https://github.com/ashnair1/COCO-Assistant/blob/master/coco_assistant/utils/det2seg.py) for getting an idea on how to convert COCO dataset to segmentation masks.

**Reply by winter2897** [reply]
Hi @ashnair1 ,
I just got back to work. Thank you for giving me advice. I am making the way you instruct. However it is experiencing this problem.
![Screenshot from 2021-03-22 08-29-18](https://user-images.githubusercontent.com/22476495/111929442-ea20dc80-8ae8-11eb-99d0-3abad0710466.png)

**Comment by winter2897** [comment]
My class: [{'supercategory': 'Dust1', 'id': 0, 'name': 'Dust1'}, {'supercategory': 'Dust2', 'id': 1, 'name': 'Dust2'}, {'supercategory': 'Dust3', 'id': 2, 'name': 'Dust3'}, {'supercategory': 'Wire', 'id': 3, 'name': 'Wire'}]. I don't have class 'supercategory': 'Background'

**Reply by ashnair1** [reply]
You don't need to have Background as a class. 

Edit: Your ids should actually start from 1.

**Comment by winter2897** [comment]
Hi @ashnair1,
```
if colour_map.shape != (len(cats) + 1, 3):
        raise AssertionError("Incorrect shape of color map array")
```
Why `len(cats) + 1`.....? Can you explain it to me?

**Reply by ashnair1** [reply]
The `+1` is for background. The palette will have `len(cats) + 1` unique colors for categories and background (which is black by default)

**Comment by winter2897** [comment, marked as answer]
I removed parameter `1` from `(len(cats) + 1, 3)`, the result showed that the label `Dust1` was the same color as Backgound. The other labels are still correctly illustrated.
![DM21-001_Sametel-031121__0000_A](https://user-images.githubusercontent.com/22476495/111942787-a341df80-8b06-11eb-8563-b0e65aa5658f.png)

**Reply by ashnair1** [reply]
You should not remove the `+1`. As i mentioned earlier, it's for background.

**Reply by ashnair1** [reply]
@winter2897 
Ah I understand the problem. It's because your category ids start from 0. Your category ids should start from 1.

Your categories should be like this:
```
[{'supercategory': 'Dust1', 'id': 1, 'name': 'Dust1'}, 
{'supercategory': 'Dust2', 'id': 2, 'name': 'Dust2'}, 
{'supercategory': 'Dust3', 'id': 3, 'name': 'Dust3'}, 
{'supercategory': 'Wire', 'id': 4, 'name': 'Wire'}]
```

I'll make a note of this in the `README.md`

**Reply by winter2897** [reply]
However, my `.json` file has the fixed definition. In your opinion how should I handle it? Instead of editing the `.json` file.

**Reply by ashnair1** [reply]
Important note: You'll need to change the category label in the annotation as well. Since the annotation might have category_id field as 0, where it should be 1.

**Reply by winter2897** [reply]
opp, okay. I will fix the `id` in the `.json` file.

**Reply by winter2897** [reply]
`Dust1` and `Wire` still missing
![DM21-001_Sametel-031121__0000_A](https://user-images.githubusercontent.com/22476495/111951722-9f1dbe00-8b16-11eb-8d07-c86fa7338335.png)

**Reply by winter2897** [reply]
"categories": [
        {
            "supercategory": "Dust1",
            "id": 1,
            "name": "Dust1"
        },
        {
            "supercategory": "Dust2",
            "id": 2,
            "name": "Dust2"
        },
        {
            "supercategory": "Dust3",
            "id": 3,
            "name": "Dust3"
        },
        {
            "supercategory": "Wire",
            "id": 4,
            "name": "Wire"
        }

**Reply by winter2897** [reply]
I fixed it. Thank you very much @ashnair1 . I have to edit "category_id" accordingly.

**Reply by winter2897** [reply]
![DM21-001_Sametel-031121__0000_A](https://user-images.githubusercontent.com/22476495/111952394-ae513b80-8b17-11eb-8cf4-2ec4ef6ad4cd.png)

**Comment by fcakyon** [comment]
@winter2897 you can use https://github.com/obss/sahi library for cropping coco annotations and images.

- Slice a coco formatted dataset:

```python
from sahi.slicing import slice_coco

coco_dict, coco_path = slice_coco(
    coco_annotation_file_path=coco_annotation_file_path,
    image_dir=image_dir,
    slice_height=256,
    slice_width=256,
    overlap_height_ratio=0.2,
    overlap_width_ratio=0.2,
)
```
